### PR TITLE
Extract SyliusUserBundle and User component

### PR DESCRIPTION
**Pull request by BartoszSiejka**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Extract SyliusUserBundle and Component.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/main.yml [hunk 1]
```diff
@@ -72,7 +72,7 @@ fos_user:
     firewall_name: main
     user_class: %sylius.model.user.class%
     group:
-        group_class: Sylius\Component\Core\Model\Group
+        group_class: Sylius\Component\User\Model\Group
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/main.yml]: Extract model class to parameter, like user class

File: src/Sylius/Bundle/UserBundle/Resources/config/validation.xml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ file that was distributed with this source code.
+
+-->
+
+<constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping"
+                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                    xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping
+                                        http://symfony.com/schema/dic/services/constraint-mapping-1.0.xsd">
+
+    <class name="Sylius\Component\Core\Model\User">
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/Resources/config/validation.xml]: ```Sylius\Component\Core\Model\User``` => ```Sylius\Component\User\Model\User```

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml [hunk 3]
```diff
@@ -100,57 +100,9 @@
...
-                <option name="groups">
-                    <value>ProfileAdd</value>
-                </option>
-            </constraint>
-        </property>
-    </class>
+    </class>    
 
-    <class name="Sylius\Component\Core\Model\Group">
+    <class name="Sylius\Component\User\Model\Group">
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml]: Extract to UserBundle

File: src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml [hunk 4]
```diff
@@ -1,34 +1,9 @@
...
-    user:
-        first_name:
-            max: First name must not be longer than {{ limit }} characters.
-            min: First name must be at least {{ limit }} characters long.
-            not_blank: Please enter your first name.
-        last_name:
-            max: Last name must not be longer than {{ limit }} characters.
-            min: Last name must be at least {{ limit }} characters long.
-            not_blank: Please enter your last name.
-    product_variant:
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml]: why is it removed?

File: src/Sylius/Bundle/UserBundle/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+CHANGELOG
+=========
+
+### v0.11.0
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/CHANGELOG.md]: v0.13.0

File: src/Sylius/Bundle/UserBundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+use Symfony\Component\Config\Definition\ConfigurationInterface;
+
+/**
+ * This class contains the configuration information for the bundle.
+ *
+ * This information is solely responsible for how the different configuration
+ * sections are normalized, and merged.
+ *
+ * @author Paweł Jędrzejewski <pjedrzejewski@diweb.pl>
```
Comment (aRn0D) [inline src/Sylius/Bundle/UserBundle/DependencyInjection/Configuration.php]: Pawel is the author of those classes ?

File: src/Sylius/Bundle/UserBundle/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        $node
+            ->children()
+                ->arrayNode('classes')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->arrayNode('user')
+                            ->addDefaultsIfNotSet()
+                            ->children()
+                                ->scalarNode('model')->defaultValue('Sylius\Component\Core\Model\User')->end()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/DependencyInjection/Configuration.php]: ```defaultValue('Sylius\Component\User\Model\User')```

File: src/Sylius/Bundle/UserBundle/Resources/config/driver/doctrine/orm.xml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services
+                               http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <parameters>
+        <parameter key="sylius.repository.group.class">Sylius\Bundle\UserBundle\Doctrine\ORM\GroupRepository</parameter>
```
Comment (aRn0D) [inline src/Sylius/Bundle/UserBundle/Resources/config/driver/doctrine/orm.xml]: It is not declare as Sylius resource ?

File: src/Sylius/Bundle/UserBundle/Templating/Helper/UserHelper.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
+<?php
+/*
```
Comment (aRn0D) [inline src/Sylius/Bundle/UserBundle/Templating/Helper/UserHelper.php]: you should remove it

File: src/Sylius/Bundle/UserBundle/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
+SyliusUserBundle [![Build status...](https://secure.travis-ci.org/Sylius/SyliusLocaleBundle.png?branch=master)](http://travis-ci.org/Sylius/SyliusLocaleBundle)
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/README.md]: ```SyliusLocaleBundle``` ?

File: src/Sylius/Bundle/UserBundle/README.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+```bash
+$ composer install
+$ bin/phpspec run -f pretty
+```
+
+Documentation
+-------------
+
+Documentation is available on [**docs.sylius.org**](http://sylius.org/en/latest/bundles/SyliusLocaleBundle/index.html).
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/README.md]: ```SyliusLocaleBundle```

File: src/Sylius/Bundle/UserBundle/README.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+Sylius twitter account
+----------------------
+
+If you want to keep up with updates, [follow the official Sylius account on twitter](http://twitter.com/Sylius).
+
+Bug tracking
+------------
+
+This bundle uses [GitHub issues](https://github.com/Sylius/SyliusLocaleBundle/issues).
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/README.md]: ``` SyliusLocaleBundle ``` ?

File: src/Sylius/Bundle/UserBundle/README.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+* New additions without breaking backwards compatibility bumps the minor.
+* Bug fixes and misc changes bump the patch.
+
+For more information on SemVer, please visit [semver.org website](http://semver.org/)...
+This versioning method is same for all **Sylius** bundles and applications.
+
+MIT License
+-----------
+
+License can be found [here](https://github.com/Sylius/SyliusLocaleBundle/blob/master/Resources/meta/LICENSE).
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/README.md]: ``` SyliusLocaleBundle ``` ?

File: src/Sylius/Bundle/UserBundle/README.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+Authors
+-------
+
+The bundle was originally created by [Paweł Jędrzejewski](http://pjedrzejewski.com).
+See the list of [contributors](https://github.com/Sylius/SyliusLocaleBundle/contributors).
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/README.md]: ``` SyliusLocaleBundle ``` ?

File: src/Sylius/Bundle/UserBundle/Resources/config/templating.xml [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services
+                               http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+
+</container>
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/Resources/config/templating.xml]: Delete empty file

File: src/Sylius/Bundle/UserBundle/Resources/translations/validators.en.yml [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    user:
+        first_name:
+            max: First name must not be longer than {{ limit }} characters.
+            min: First name must be at least {{ limit }} characters long.
+            not_blank: Please enter your first name.
+        last_name:
+            max: Last name must not be longer than {{ limit }} characters.
+            min: Last name must be at least {{ limit }} characters long.
+            not_blank: Please enter your last name.
+    product_variant:
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/Resources/translations/validators.en.yml]: why is it here?

File: src/Sylius/Bundle/UserBundle/Templating/Helper/UserHelper.php [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+//    }
+//
+//    /**
+//     * {@inheritdoc}
+//     */
+//    public function getName()
+//    {
+//        return 'sylius_locale';
+//    }
+//}
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/Templating/Helper/UserHelper.php]: Delete file

File: src/Sylius/Bundle/UserBundle/composer.json [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    "description": "Users management for Symfony2 projects.",
+    "keywords": ["language", "localization", "user", "i18n", "internationalization"],
+    "homepage": "http://sylius.org",
+    "license": "MIT",
+    "authors": [
+        {
+            "name":     "Paweł Jędrzejewski",
+            "homepage": "http://pjedrzejewski.com"
+        },
+        {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/composer.json]: Your name here

File: src/Sylius/Component/Core/Model/User.php [hunk 19]
```diff
@@ -237,85 +152,4 @@ public function isDeleted()
     {
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/User.php]: Extract to User component and add to interface

File: src/Sylius/Component/Core/spec/Model/GroupSpec.php [hunk 20]
```diff
@@ -20,12 +20,12 @@ class GroupSpec extends ObjectBehavior
...
+        $this->shouldHaveType('Sylius\Component\User\Model\Group');
     }
 
     function it_implements_Sylius_group_interface()
     {
-        $this->shouldImplement('Sylius\Component\Core\Model\GroupInterface');
+        $this->shouldImplement('Sylius\Component\User\Model\GroupInterface');
     }
 
     function it_extends_FOS_group_model()
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/GroupSpec.php]: Move spec to User component

File: src/Sylius/Component/Core/spec/Model/UserSpec.php [hunk 21]
```diff
@@ -25,9 +25,9 @@ function it_is_initializable()
         $this->shouldHaveType('Sylius\Component\Core\Model\User');
     }
 
-    function it_implements_Fos_user_interface()
+    function it_implements_user_component_interface()
     {
-        $this->shouldImplement('FOS\UserBundle\Model\UserInterface');
+        $this->shouldImplement('Sylius\Component\User\Model\UserInterface');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/UserSpec.php]: As it it Core user it should implement ``` Sylius\Component\Core\Model\UserInterface ```

File: src/Sylius/Component/User/CHANGELOG.md [hunk 22]
```diff
@@ -0,0 +1,6 @@
+CHANGELOG
+=========
+
+### v0.11.0
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/CHANGELOG.md]: v0.13.0

File: src/Sylius/Component/User/Model/User.php [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    protected $createdAt;
+    protected $updatedAt;
+    protected $deletedAt;
+    
+    public function __construct()
+    {
+        $this->createdAt     = new \DateTime();
+        $this->oauthAccounts = new ArrayCollection();
+
+        parent::__construct();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/Model/User.php]: Parent constructor should be called first

File: src/Sylius/Component/User/README.md [hunk 24]
```diff
@@ -0,0 +1,47 @@
+User Component [![Build status...](https://secure.travis-ci.org/Sylius/Locale.png?branch=master)](http://travis-ci.org/Sylius/Locale)
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/README.md]: ``` Locale ``` ? Correct all below also

File: src/Sylius/Component/User/composer.json [hunk 25]
```diff
@@ -0,0 +1,45 @@
+{
+    "name": "sylius/user",
+    "type": "library",
+    "description": "Locale handling for PHP applications.",
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/composer.json]: Locale handling ?

File: src/Sylius/Component/User/composer.json [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+    "keywords": ["language", "localization", "locale", "i18n", "internationalization"],
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/composer.json]: Correct these keywords

File: src/Sylius/Component/User/spec/Model/UserSpec.php [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+{
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Component\User\Model\User');
+    }
+    
+    function it_implements_Fos_user_interface()
+    {
+        $this->shouldImplement('FOS\UserBundle\Model\UserInterface');
+    }
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/spec/Model/UserSpec.php]: Add some more specs

File: src/Sylius/Bundle/UserBundle/composer.json [hunk 28]
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "sylius/user-bundle",
+    "type": "symfony-bundle",
+    "description": "Users management for Symfony2 projects.",
+    "keywords": ["language", "localization", "user", "i18n", "internationalization"],
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/composer.json]: Correct keywords

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml [hunk 29]
```diff
@@ -184,6 +184,23 @@ sylius_sequence:
     generators:
         %sylius.model.order.class%: sylius.sequence.sequential_number_generator
 
+sylius_resource:
+    resources:
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Extract to UserBundle

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml [hunk 30]
```diff
@@ -246,4 +263,6 @@ sylius_translation:
             fallbackLocale: fallbackLocale
         translation:
             field: translatable
-            locale: locale
\ No newline at end of file
+            locale: locale
+            
+sylius_user: ~
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Configure Core model and controller here

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml [hunk 31]
```diff
@@ -184,6 +184,23 @@ sylius_sequence:
...
         %sylius.model.order.class%: sylius.sequence.sequential_number_generator
 
+sylius_resource:
+    resources:
+        sylius.user:
+            driver: doctrine/orm
+            classes:
+                model: Sylius\Component\Core\Model\User
+                controller: Sylius\Bundle\CoreBundle\Controller\UserController
+                repository: Sylius\Bundle\UserBundle\Doctrine\ORM\UserRepository
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: remove repository (defined in parameter)

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml (hunk 29 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml]: Remove this

File: src/Sylius/Bundle/UserBundle/Resources/config/driver/doctrine/orm.xml (hunk 8 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UserBundle/Resources/config/driver/doctrine/orm.xml]: Custom repository is defined here, Sylius resource uses default EntityRepository

File: src/Sylius/Component/User/spec/Model/UserSpec.php [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        $deletedAt = new \DateTime('yesterday');
+        $this->setDeletedAt($deletedAt);
+        
+        $this->isDeleted()->shouldReturn(true);
+    }
+    
+    function it_should_return_false_if_user_is_not_deleted()
+    {        
+        $this->isDeleted()->shouldReturn(false);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/spec/Model/UserSpec.php]: $this->shouldNotBeDeleted();

File: src/Sylius/Component/User/spec/Model/UserSpec.php [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+        
+        $this->getFullName()->shouldReturn('Edward Kenway');
+    }
+    
+    function it_should_return_true_if_user_is_deleted()
+    {
+        $deletedAt = new \DateTime('yesterday');
+        $this->setDeletedAt($deletedAt);
+        
+        $this->isDeleted()->shouldReturn(true);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/spec/Model/UserSpec.php]: $this->shouldBeDeleted();

File: src/Sylius/Component/User/spec/Model/UserSpec.php [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+    {        
+        $this->isDeleted()->shouldReturn(false);
+    }
+    
+    function it_should_return_false_if_user_deleted_time_is_future_date()
+    {
+        $deletedAt = new \DateTime('tomorrow');
+        $this->setDeletedAt($deletedAt);
+        
+        $this->isDeleted()->shouldReturn(false);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/User/spec/Model/UserSpec.php]: $this->shouldNotBeDeleted();

**Comment by aRn0D** [comment]
@BartoszSiejka What is the status of that? I think you should squash some commit like Add some user specs and Fixed specs ;)

**Comment by aRn0D** [comment]
@BartoszSiejka ping !

**Comment by aramalipoor** [comment]
@BartoszSiejka pong!

**Comment by michalmarcinkowski** [comment]
@aRn0D @aramalipoor I took it over and will make a PR together with the customer-user split in the coming days.

**Pull request closed without merging**